### ISSUE TITLE
Mock Stubbing and Monkey Patching

### ISSUE THREAD

**Issue by MrYsLab** [pytest-dev/pytest-asyncio#7]
Does mock monkey patching work with pytest-dev?

I have 3 similar tests that allow me to stub out an instantiation of a pyserial object. The mock stub works in all 3 tests I have included below. However, I am also trying to monkeypatch a call to the get() method on a queue.

The first test succeeds, but I am doing a direct call to get and not yield from .

The second test is essentially the same as the first, but I use a yield from.

The third test just changes the return statement to a yield in the mockreturn function.

Below are the 3 test cases and the outputs from the failures.

Thanks,
Alan
```
import asyncio
import pytest
import serial

from esplora_serial import EsploraSerial


def test_some_interaction(monkeypatch, mocker):
    # create a queue
    q = asyncio.LifoQueue(maxsize=1)

    # stub out the call to serial.Serial in the EsploraSerial class __init__
    mocker.patch('serial.Serial')

    # instantiate the Esplora Serial class
    my_serial = EsploraSerial("/dev/ttyACM0", q)

    #create a mock return for the queue "get"
    def mockreturn(get):
        return '/abc'

    # monkey patch the lifo get to return "/abc"
    monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)

    resp = q.get()
    assert resp == "/abc"


@pytest.mark.asyncio
def test_some_interaction2(monkeypatch, mocker):
    # create a queue
    q = asyncio.LifoQueue(maxsize=1)

    # stub out the call to serial.Serial in the EsploraSerial class __init__
    mocker.patch('serial.Serial')

    # instantiate the Esplora Serial class
    my_serial = EsploraSerial("/dev/ttyACM0", q)

    #create a mock return for the queue "get"
    def mockreturn(get):
        return '/abc'

    # monkey patch the lifo get to return "/abc"
    monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)

    resp = yield from  q.get()
    assert resp == "/abc"

@pytest.mark.asyncio
def test_some_interaction3(monkeypatch, mocker):
    # create a queue
    q = asyncio.LifoQueue(maxsize=1)

    # stub out the call to serial.Serial in the EsploraSerial class __init__
    mocker.patch('serial.Serial')

    # instantiate the Esplora Serial class
    my_serial = EsploraSerial("/dev/ttyACM0", q)

    #create a mock return for the queue "get"
    def mockreturn(get):
        yield '/abc'

    # monkey patch the lifo get to return "/abc"
    monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)

    resp = yield from  q.get()
    assert resp == "/abc"
```

And now the test output:
```
/usr/local/bin/python3 /home/afy/PycharmProjects/pyvmmonitor/public_api/pyvmmonitor/__init__.py --profile=lsprof --spawn-ui=false /home/afy/pycharm-4.0.4/helpers/pycharm/pytestrunner.py -p pytest_teamcity /home/afy/PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py
Testing started at 7:26 PM ...
Patching args: ['/bin/sh', '-c', 'uname -p 2> /dev/null']
Process is not python, returning.
============================= test session starts ==============================
platform linux -- Python 3.4.3 -- py-1.4.26 -- pytest-2.7.0
rootdir: /home/afy/PycharmProjects/esp4s-aio/test, inifile: 
plugins: asyncio, mock
collected 3 items

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py .F
monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0xb4b146cc>
mocker = <pytest_mock.MockFixture object at 0xb4b1478c>

    @pytest.mark.asyncio
    def test_some_interaction2(monkeypatch, mocker):
        # create a queue
        q = asyncio.LifoQueue(maxsize=1)
    
        # stub out the call to serial.Serial in the EsploraSerial class __init__
        mocker.patch('serial.Serial')
    
        # instantiate the Esplora Serial class
        my_serial = EsploraSerial("/dev/ttyACM0", q)
    
        #create a mock return for the queue "get"
        def mockreturn(get):
            return '/abc'
    
        # monkey patch the lifo get to return "/abc"
        monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)
    
>       resp = yield from  q.get()
E       RuntimeError: Task got bad yield: '/'

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:115: RuntimeError
F
monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0xb4b6156c>
mocker = <pytest_mock.MockFixture object at 0xb4b616cc>

    @pytest.mark.asyncio
    def test_some_interaction3(monkeypatch, mocker):
        # create a queue
        q = asyncio.LifoQueue(maxsize=1)
    
        # stub out the call to serial.Serial in the EsploraSerial class __init__
        mocker.patch('serial.Serial')
    
        # instantiate the Esplora Serial class
        my_serial = EsploraSerial("/dev/ttyACM0", q)
    
        #create a mock return for the queue "get"
        def mockreturn(get):
            yield '/abc'
    
        # monkey patch the lifo get to return "/abc"
        monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)
    
>       resp = yield from  q.get()

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:136: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

get = <LifoQueue at 0xb4b615ac maxsize=1>

    def mockreturn(get):
>       yield '/abc'
E       RuntimeError: Task got bad yield: '/abc'

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:131: RuntimeError


=================================== FAILURES ===================================
____________________________ test_some_interaction2 ____________________________

monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0xb4b146cc>
mocker = <pytest_mock.MockFixture object at 0xb4b1478c>

    @pytest.mark.asyncio
    def test_some_interaction2(monkeypatch, mocker):
        # create a queue
        q = asyncio.LifoQueue(maxsize=1)
    
        # stub out the call to serial.Serial in the EsploraSerial class __init__
        mocker.patch('serial.Serial')
    
        # instantiate the Esplora Serial class
        my_serial = EsploraSerial("/dev/ttyACM0", q)
    
        #create a mock return for the queue "get"
        def mockreturn(get):
            return '/abc'
    
        # monkey patch the lifo get to return "/abc"
        monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)
    
>       resp = yield from  q.get()
E       RuntimeError: Task got bad yield: '/'

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:115: RuntimeError
____________________________ test_some_interaction3 ____________________________

monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0xb4b6156c>
mocker = <pytest_mock.MockFixture object at 0xb4b616cc>

    @pytest.mark.asyncio
    def test_some_interaction3(monkeypatch, mocker):
        # create a queue
        q = asyncio.LifoQueue(maxsize=1)
    
        # stub out the call to serial.Serial in the EsploraSerial class __init__
        mocker.patch('serial.Serial')
    
        # instantiate the Esplora Serial class
        my_serial = EsploraSerial("/dev/ttyACM0", q)
    
        #create a mock return for the queue "get"
        def mockreturn(get):
            yield '/abc'
    
        # monkey patch the lifo get to return "/abc"
        monkeypatch.setattr(asyncio.LifoQueue, 'get', mockreturn)
    
>       resp = yield from  q.get()

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:136: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

get = <LifoQueue at 0xb4b615ac maxsize=1>

    def mockreturn(get):
>       yield '/abc'
E       RuntimeError: Task got bad yield: '/abc'

PycharmProjects/esp4s-aio/test/test_esp4sHttpServer.py:131: RuntimeError
====================== 2 failed, 1 passed in 0.03 seconds ======================

Process finished with exit code 0
```

**Comment by Tinche** [comment]
Hello,

yes, monkey patching works with pytest-asyncio; it works by default and we don't do anything to stop it.

Your problem is you're monkey patching coroutines with ordinary functions. I don't know what exactly EsploraSerial is, so I've removed it, but here's a gist with working examples: https://gist.github.com/Tinche/dcfe90292b8fb94f89e7 (I've added @asyncio.coroutine decorators to your mock functions to make them into coroutines, since the default AsyncQueue.get() is a coroutine itself, and replaced the yield in the third function with the proper return).
